### PR TITLE
WIP: Adds basic features

### PR DESCRIPTION
**Pull request by daften**
Provides the basic framework:
- README
- composer.json
- Travis testing

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+language: php
+php:
+  - '7'
+  - 5.6
+  - 5.5
+install:
+  - composer self-update
+  - composer install
+script:
+  - vendor/bin/robo test
```
Comment (Jelle-S) [inline .travis.yml]: Is an empty newline at the end of a yaml file a requirement as well? If it is, it's missing here.

File: Robofile.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<?php
+class RoboFile extends \Robo\Tasks
```
Comment (Jelle-S) [inline Robofile.php]: Newline after opening tag if we want to follow Drupal coding standards.

File: Robofile.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    $this->taskPHPUnit()
+      ->option('disallow-test-output')
+      ->option('report-useless-tests')
+      ->option('strict-coverage')
+      ->option('-v')
+      ->option('-d error_reporting=-1')
+      ->arg('tests')
+      ->run();
+  }
+}
```
Comment (Jelle-S) [inline Robofile.php]: Newline at end of file.

File: composer.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    "type": "robo-tasks",
+    "require": {
+        "consolidation/robo": "~1.0",
+        "drupal/console": "~1.0",
+        "php": ">=5.5.0"
+    },
+    "require-dev": {
+        "phpunit/phpunit": "~4.4"
+    },
+    "minimum-stability": "dev",
```
Comment (Jelle-S) [inline composer.json]: Why not stable?

File: tests/DrupalConsoleStackTest.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  public function testDrupalConsoleStatus()
+  {
+    $result = $this->taskDrupalConsoleStack(__DIR__ . '/../vendor/bin/drupal')
+      ->printed(false)
+      ->status()
+      ->run();
+    $this->assertTrue($result->wasSuccessful(), 'Exit code was: ' . $result->getExitCode());
+  }
+
+}
```
Comment (Jelle-S) [inline tests/DrupalConsoleStackTest.php]: Newline at end of file.

File: composer.json (hunk 4 above)
Comment (daften) [reply composer.json]: Was voor een dependency van een dependency ...

File: Robofile.php (hunk 2 above)
Comment (daften) [reply Robofile.php]: PSR-2 volgen, niet Drupal coding standards (IMO)

File: .scrutinizer.yml [hunk 6]
```diff
@@ -0,0 +1,44 @@
+filter:
+    excluded_paths:
+        - tests/
+        - vendor/
+before_commands:
+    - composer install
+tools:
+    php_code_sniffer:
+        config:
+            standard: PSR2
```
Comment (Jelle-S) [inline .scrutinizer.yml]: PSR-12 is an option as well: https://github.com/php-fig/fig-standards/blob/master/proposed/extended-coding-style-guide.md
Thoughts?

File: src/DrupalConsoleStack.php [hunk 7]
```diff
@@ -0,0 +1,698 @@
+<?php
+
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php:2]: Remove newline.

File: src/DrupalConsoleStack.php [hunk 8]
```diff
@@ -0,0 +1,698 @@
...
+     *   Whether or not the maintenance mode should be on or off.
+     *
+     * @return $this
+     */
+    public function maintenance($mode = true)
+    {
+        $maintenanceMode = $mode ? 'on' : 'off';
+        $this->printTaskInfo("Set maintenance mode $maintenanceMode");
+
+        return $this->drupal('site:maintenance $maintenanceMode');
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php]: Either concatenate or use double quotes.

File: src/DrupalConsoleStack.php [hunk 9]
```diff
@@ -0,0 +1,698 @@
...
+     * Executes `drupal site:install`.
+     *
+     * @param string $installationProfile
+     *   The installation profile to use during install.
+     *
+     * @return $this
+     */
+    public function siteInstall($installationProfile = '')
+    {
+        return $this->drupal('site:install ' . $installationProfile);
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php]: We should be consistent, either use concatenation everywhere (my personal preference, but it seems the majority is the other way round in this file) or use double quotes with inline variables.

File: src/loadTasks.php [hunk 10]
```diff
@@ -0,0 +1,22 @@
+<?php
+
```
Comment (Jelle-S) [inline src/loadTasks.php:2]: Remove newline.

File: tests/DrupalConsoleStackTest.php [hunk 11]
```diff
@@ -0,0 +1,141 @@
+<?php
+
```
Comment (Jelle-S) [inline tests/DrupalConsoleStackTest.php:2]: Remove newline.

File: tests/DrupalConsoleStackTest.php [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+use Robo\Robo;
+
+class DrupalConsoleStackTest extends \PHPUnit_Framework_TestCase implements ContainerAwareInterface
+{
+
+    use DigipolisGent\Robo\Task\DrupalConsole\loadTasks;
+    use TaskAccessor;
+    use ContainerAwareTrait;
+
+    // Set up the Robo container so that we can create tasks in our tests.
```
Comment (Jelle-S) [inline tests/DrupalConsoleStackTest.php]: Use proper docblock.

File: tests/DrupalConsoleStackTest.php [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+    public function setup()
+    {
+        $container = Robo::createDefaultContainer(null, new NullOutput());
+        $this->setContainer($container);
+    }
+
+    // Scaffold the collection builder
```
Comment (Jelle-S) [inline tests/DrupalConsoleStackTest.php]: Use proper docblock.

File: tests/DrupalConsoleStackTest.php [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+          ->getCommand();
+        $expected = 'drupal site:install standard --yes --site-name=' . escapeshellarg('Site Name Mysql')
+          . ' --site-mail=site-mail2@example.com --langcode=fr'
+          . ' --account-mail=mail2@example.com --account-name=' . escapeshellarg('admin-user')
+          . ' --account-pass=passw --db-prefix=drupal_ --db-type=mysql'
+          . ' --db-host=' . escapeshellarg('localhost') . ' --db-name=' . escapeshellarg('testdb')
+          . ' --db-user=' . escapeshellarg('dbuser') . ' --db-pass=' . escapeshellarg('testdbpw');
+        $this->assertEquals($expected, $command);
+    }
+
```
Comment (Jelle-S) [inline tests/DrupalConsoleStackTest.php]: This and all methods above: Add docblocks.

File: .scrutinizer.yml [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+            paths: ["*.php"]
+    php_cpd:
+        min_lines: 5
+        min_tokens: 70
+        names:
+            - '*.php'
+        excluded_dirs:
+            - tests/
+            - vendor/
+        enabled: false
```
Comment (Jelle-S) [inline .scrutinizer.yml]: Why not enabled?

File: RoboFile.php [hunk 16]
```diff
@@ -0,0 +1,21 @@
+<?php
+
```
Comment (Jelle-S) [inline RoboFile.php:2]: Remove newline.

Also: This file needs docblocks.

File: src/DrupalConsoleStack.php [hunk 17]
```diff
@@ -0,0 +1,698 @@
...
+    /**
+     * Verbosity levels:
+     * 1 for normal output, 2 for more verbose output, and 3 for debug.
+     */
+    const VERBOSITY_LEVEL_NORMAL = 1;
+    const VERBOSITY_LEVEL_VERBOSE = 2;
+    const VERBOSITY_LEVEL_DEBUG = 3;
+
+    /**
+     * Pass arguments to the executable; applies to the next command only.
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php]: "Arguments to pass to the executable." sounds better here.

File: src/DrupalConsoleStack.php [hunk 18]
```diff
@@ -0,0 +1,698 @@
...
+     *
+     * @return $this
+     */
+    protected function argForNextCommand($arg)
+    {
+        return $this->argsForNextCommand($arg);
+    }
+
+    /**
+     * Pass methods parameters as arguments to executable;
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php]: s/methods/method's/ (posessive 's).

File: src/DrupalConsoleStack.php [hunk 19]
```diff
@@ -0,0 +1,698 @@
...
+     * Sets the site name.
+     *
+     * @param string $siteName
+     *   The site name.
+     *
+     * @return $this
+     */
+    public function siteName($siteName)
+    {
+        $this->argForNextCommand('--site-name=' . escapeshellarg($siteName));
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php]: We don't seem to be consistent in when we escape the argument and when we don't. Maybe we should split up the arguments for argForNextCommand in to $option and $value. Do something like:

```
protected function argForNextCommand($option, $value = NULL)
{
    return $this->argsForNextCommand([[$option => $value]]);
}

protected function argsForNextCommand($args)
{
    foreach ($args as $arg) {
        foreach ($arg as $option => $value) {
            $this->argumentsForNextCommand .= ' --' . $option . (is_null($value) ? '' : '=' . escapeshellarg($value));
        }
    return $this;
}
```

File: src/DrupalConsoleStack.php [hunk 20]
```diff
@@ -0,0 +1,698 @@
...
+     *
+     * @return $this
+     */
+    public function siteInstall($installationProfile = '')
+    {
+        return $this->drupal('site:install ' . $installationProfile);
+    }
+
+    /**
+     * Export configuration
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php]: Punctuation.

File: src/DrupalConsoleStack.php [hunk 21]
```diff
@@ -0,0 +1,698 @@
...
+    public function configExport()
+    {
+        $this->printTaskInfo('Export configuration');
+        $this->drupal('config:export');
+
+        return $this;
+    }
+
+    /**
+     * Import configuration
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php]: Punctuation.

File: src/DrupalConsoleStack.php [hunk 22]
```diff
@@ -0,0 +1,698 @@
...
+    public function dbDrop($database)
+    {
+        $this->printTaskInfo("Drops database $database");
+        $this->drupal("database:drop $database");
+
+        return $this;
+    }
+
+    /**
+     * Drops the database.
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php]: Copy paste mistake. Update the description.

File: src/DrupalConsoleStack.php [hunk 23]
```diff
@@ -0,0 +1,698 @@
...
+    {
+        $migrationIdsString = implode(',', $migrationIds);
+        $this->printTaskInfo("Execute migrations $migrationIdsString");
+        $this->drupal("migrate:execute $migrationIdsString");
+
+        return $this;
+    }
+
+    /**
+     * Executes `drupal list`
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php]: Punctuation.

File: src/DrupalConsoleStack.php [hunk 24]
```diff
@@ -0,0 +1,698 @@
...
+     *
+     * @return $this
+     */
+    public function listCommands()
+    {
+        return $this->drupal('list');
+    }
+
+    /**
+     * Executes `drupal site:status`
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php]: Punctuation.

File: tests/DrupalConsoleStackTest.php [hunk 25]
```diff
@@ -0,0 +1,141 @@
+<?php
+
+use League\Container\ContainerAwareInterface;
+use League\Container\ContainerAwareTrait;
+use Symfony\Component\Console\Output\NullOutput;
```
Comment (Jelle-S) [inline tests/DrupalConsoleStackTest.php]: Personal preference: Sort use statements alphabetically.

File: tests/DrupalConsoleStackTest.php [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    public function testDrupalConsoleList()
+    {
+        $result = $this->taskDrupalConsoleStack(__DIR__ . '/../vendor/bin/drupal')
+          ->printed(false)
+          ->listCommands()
+          ->run();
+        $this->assertEquals(trim($result->getMessage()),
+          'DrupalConsole must be executed within a Drupal Site.');
```
Comment (Jelle-S) [inline tests/DrupalConsoleStackTest.php]: You can execute `drupal list` outside of a Drupal site. See also [this comment](https://github.com/digipolisgent/robo-drupal-console/commit/d571633a276785f0cc61c2e0a57afd9b1ca49090#commitcomment-20181746).

File: .scrutinizer.yml (hunk 15 above)
Comment (Jelle-S) [reply .scrutinizer.yml]: Since we are probably going to switch to codeclimate, this file should be removed.
@daften ?

File: .travis.yml [hunk 27]
```diff
@@ -0,0 +1,13 @@
...
+  - 5.6
+  - 5.5
+matrix:
+  allow_failures:
+    - php: 7.0 # This should be fixed, temporary for the PHP7 support badge
+install:
+  - composer self-update
+  - composer install
+script:
+  - vendor/bin/robo test
```
Comment (Jelle-S) [inline .travis.yml:11]: we should probably add https://docs.codeclimate.com/v1.0/docs/travis-ci-php-test-coverage once we make the switch to codeclimate.

File: composer.json [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+    "autoload": {
+        "psr-4": {
+            "DigipolisGent\\Robo\\Task\\DrupalConsole\\": "src"
```
Comment (Jelle-S) [inline composer.json:16]: Should we namespace our tests as well?

File: composer.json [hunk 29]
```diff
@@ -0,0 +1,29 @@
...
+    },
+    "scripts": {
+        "test": "robo --ansi test"
+    },
+    "license": "MIT",
+    "authors": [
+        {
+            "name": "Dieter Blomme",
+            "email": "dieterblomme@gmail.com"
+        }
```
Comment (Jelle-S) [inline composer.json:36]: I should maybe add myself here one day... :)

File: src/DrupalConsoleStack.php [hunk 30]
```diff
@@ -0,0 +1,712 @@
...
+     * Pass an option to the executable; applies to the next command only.
+     *
+     * @param string $name
+     *   The name of the option to pass.
+     * @param string $value
+     *   The value for this option to pass (optional).
+     *
+     * @return $this
+     */
+    protected function optionForNextCommand($name, $value = NULL)
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php]: s/NULL/null/

File: src/DrupalConsoleStack.php [hunk 31]
```diff
@@ -0,0 +1,712 @@
...
+    {
+        $this->executable = $pathToDrupalConsole;
+    }
+
+    /**
+     * Pass an option to the executable; applies to the next command only.
+     *
+     * @param string $name
+     *   The name of the option to pass.
+     * @param string $value
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php]: s/string/string|null/

File: src/DrupalConsoleStack.php [hunk 32]
```diff
@@ -0,0 +1,712 @@
...
+     */
+    public function cacheRebuild($cacheName = 'all')
+    {
+        $this->printTaskInfo('Cache rebuild');
+
+        return $this->drupal('cache:rebuild ' . static::escape($cacheName));
+    }
+
+    /**
+     * Execute a specific Update N function in a module, or execute all.
```
Comment (Jelle-S) [inline src/DrupalConsoleStack.php]: s/Update/update/

File: tests/DrupalConsoleStackTest.php [hunk 33]
```diff
@@ -0,0 +1,405 @@
...
+
+    /**
+     * Test the db file option.
+     */
+    public function testDbFileOption() {
+        $this->doTestOption('dbFile', 'db-file', '.h.sqlite');
+    }
+
+    /**
+     * Test the db file option.
```
Comment (Jelle-S) [inline tests/DrupalConsoleStackTest.php]: s/file/host/

File: tests/DrupalConsoleStackTest.php [hunk 34]
```diff
@@ -0,0 +1,405 @@
...
+
+    /**
+     * Test the db prefix option.
+     */
+    public function testDbPrefixOption() {
+        $this->doTestOption('dbPrefix', 'db-prefix', 'prefix_');
+    }
+
+    /**
+     * Test the db password option.
```
Comment (Jelle-S) [inline tests/DrupalConsoleStackTest.php]: s/password/port/

File: tests/DrupalConsoleStackTest.php [hunk 35]
```diff
@@ -0,0 +1,405 @@
...
+     * Test the account name option.
+     */
+    public function testAccountNameOption() {
+        $this->doTestOption('accountName', 'account-name', 'accountName');
+    }
+
+    /**
+     * Test the account password option.
+     */
+    public function testAccountpassOption() {
```
Comment (Jelle-S) [inline tests/DrupalConsoleStackTest.php]: s/testAccountpassOption/testAccountPassOption

File: .scrutinizer.yml (hunk 15 above)
Comment (daften) [reply .scrutinizer.yml]: Disabled want copy/paste.
Mag hier weg, maar wil scrutinizer wel nog verder evalueren, dat zal dan op een aparte branch zijn, maak die mss eerst (feature/scrutinizer) en push die aub

File: .travis.yml (hunk 27 above)
Comment (daften) [reply .travis.yml:11]: Kijk eens op de codeclimate branch, is daar normaal gebeurd

File: composer.json (hunk 28 above)
Comment (daften) [reply composer.json:16]: Is volgens mij geen standard practice om dat te doen, autoload is voor packages die deze includen, testen includen is nu niet echt nodig IMO.
@zero2one thoughts?

File: composer.json (hunk 29 above)
Comment (daften) [reply composer.json:36]: Today is that day!

File: .travis.yml (hunk 27 above)
Comment (Jelle-S) [reply .travis.yml:11]: Indeed

File: .editorconfig [hunk 36]
```diff
@@ -0,0 +1,14 @@
+# This file is for unifying the coding style for different editors and IDEs
+# editorconfig.org
+
+root = true
+
+[*]
+end_of_line = lf
+charset = utf-8
```
Comment (zero2one) [inline .editorconfig:8]: Add forcing spaces for intends:

```
indent_style = space
```

File: .editorconfig [hunk 37]
```diff
@@ -0,0 +1,14 @@
...
+trim_trailing_whitespace = true
+insert_final_newline = true
+
+[**.php]
+indent_style = space
+indent_size = 4
```
Comment (zero2one) [inline .editorconfig:15]: Add support for markdown files.

```
[*.md]
trim_trailing_whitespace = false
```

File: .travis.yml [hunk 38]
```diff
@@ -0,0 +1,13 @@
+language: php
+php:
+  - 7.0
```
Comment (zero2one) [inline .travis.yml:4]: Include also 7.1

File: composer.json (hunk 28 above)
Comment (zero2one) [reply composer.json:16]: You should namespace your tests to.

```json
    "autoload": {
        "psr-4": {
            "DigipolisGent\\Robo\\Task\\DrupalConsole\\": "src"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "DigipolisGent\\Tests\\Robo\\Task\\DrupalConsole\\": "tests"
        }
    },
```

File: tests/DrupalConsoleStackTest.php [hunk 39]
```diff
@@ -0,0 +1,405 @@
+<?php
+use DigipolisGent\Robo\Task\DrupalConsole\DrupalConsoleStack;
```
Comment (zero2one) [inline tests/DrupalConsoleStackTest.php:5]: Add namespace:

```php
namespace DigipolisGent\Tests\Robo\Task\DrupalConsole;
```

File: tests/DrupalConsoleStackTest.php [hunk 40]
```diff
@@ -0,0 +1,405 @@
...
+
+    /**
+     * Scaffold the collection builder.
+     *
+     * @return \Robo\Collection\CollectionBuilder
+     *   The collection builder.
+     */
+    public function collectionBuilder()
+    {
+        $emptyRobofile = new \Robo\Tasks;
```
Comment (zero2one) [inline tests/DrupalConsoleStackTest.php]: Initiate classes always including the ();

```php
$emptyRobofile = new \Robo\Tasks();
```

File: tests/DrupalConsoleStackTest.php [hunk 41]
```diff
@@ -0,0 +1,405 @@
...
+     * @param string $command
+     *   The expected command, without the 'drupal ' prefix and ' --yes' suffix.
+     * @param array $args
+     */
+    protected function doTestCommand($method, $command, array $args = [])
+    {
+        $stack = $this->taskDrupalConsoleStack();
+        call_user_func_array([$stack, $method], $args);
+        $expected = 'drupal ' . $command . ' --yes';
+        $this->assertEquals($expected, $stack->getCommand());
```
Comment (zero2one) [inline tests/DrupalConsoleStackTest.php]: Don't hide tests in private methods, move the actual check to the `testMethod()`.

```php
protected function doTestCommand($method, $command, array $args = [])
{
    $stack = $this->taskDrupalConsoleStack();
    call_user_func_array([$stack, $method], $args);
    return $stack->getCommand();
}

function testCacheRebuildCommand()
{
    $this->assertEquals(
        'drupal cache:rebuild all --yes',
        $this->doTestCommand('cacheRebuild', 'cache:rebuild ' . static::escape('all'))
    );

$this->assertEquals(
        'drupal cache:rebuild menu --yes',
        $this->doTestCommand('cacheRebuild', 'cache:rebuild ' . static::escape('menu'), ['menu'])
    );
}
```

File: tests/DrupalConsoleStackTest.php [hunk 42]
```diff
@@ -0,0 +1,405 @@
...
+     * @param string $method
+     *   The method to call on DrupalConsoleStack.
+     * @param string $option
+     *   The option name.
+     * @param string $value
+     *   The option value.
+     */
+    protected function doTestOption($method, $option, $value = null)
+    {
+        $stack = $this->taskDrupalConsoleStack();
```
Comment (zero2one) [inline tests/DrupalConsoleStackTest.php:73]: Don't hide tests in private methods, move the actual check to the testMethod().

File: tests/DrupalConsoleStackTest.php [hunk 43]
```diff
@@ -0,0 +1,405 @@
...
+     * Test if options are prepended before each command.
+     */
+    public function testOptionsArePrependedBeforeEachCommand()
+    {
+        $command = $this->taskDrupalConsoleStack()
+          ->drupalRootDirectory('/var/www/html/app')
+          ->drupal('command-1')
+          ->drupal('command-2')
+          ->getCommand();
+        $this->assertEquals(2, preg_match_all('#--root /var/www/html/app#', $command));
```
Comment (zero2one) [inline tests/DrupalConsoleStackTest.php]: use `$this->assertRegExp() `

See https://phpunit.de/manual/current/en/appendixes.assertions.html

File: RoboFile.php [hunk 44]
```diff
@@ -0,0 +1,39 @@
...
+    use DigipolisGent\Robo\Task\DrupalConsole\loadTasks;
+
+    /**
+     * Runs the unit tests.
+     */
+    public function test()
+    {
+        $this->stopOnFail(true);
+        $this->taskPHPUnit()
+          ->option('disallow-test-output')
```
Comment (zero2one) [inline RoboFile.php]: Intend should be 4 spaces.

Lines with 3 tabs should have 12 spaces in front…

```php
    public function test()
    {
        $this->stopOnFail(true);
        $this->taskPHPUnit()
            ->option('disallow-test-output')
            ->option('report-useless-tests')
            ->option('strict-coverage')
            ->option('-v')
            ->option('-d error_reporting=-1')
            ->arg('tests')
            ->run();
    }
```

And if you want to have better allignment:

```php
    public function test()
    {
        $this->stopOnFail(true);
        $this
            ->taskPHPUnit()
            ->option('disallow-test-output')
            ->option('report-useless-tests')
            ->option('strict-coverage')
            ->option('-v')
            ->option('-d error_reporting=-1')
            ->arg('tests')
            ->run();
    }
```

**Review by zero2one — APPROVED** [review]

**Pull request merged**